sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,1061 @@
+      });
+}
+
+/**
+ * Mark the end of a buffer write operation for an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ */
+static void buffer_op_stop_write(dif_usbdev_t *usbdev, uint8_t endpoint) {
+  usbdev->active_buffer_ops = bitfield_set_field32(
+      usbdev->active_buffer_ops,
+      (bitfield_field32_t){
+          .mask = 1, .index = kEndpointHwInfos[endpoint].bit_index, .value = 0,
+      });
+}
+
+/**
+ * Check if there is an active buffer write operation for an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ * @return `true` if there is an active buffer write operation for the endpoint,
+ * `false` otherwise.
+ */
+static bool buffer_op_is_writing(dif_usbdev_t *usbdev, uint8_t endpoint) {
+  return bitfield_get_field32(
+             usbdev->active_buffer_ops,
+             (bitfield_field32_t){
+                 .mask = 1, .index = kEndpointHwInfos[endpoint].bit_index,
+             }) > 0;
+}
+
+/**
+ * Mark the start of a buffer read operation for an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ */
+static void buffer_op_start_read(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops = bitfield_set_field32(
+      usbdev->active_buffer_ops,
+      (bitfield_field32_t){
+          .mask = 1, .index = USBDEV_READ_BUFFER_OP_BIT_INDEX, .value = 1,
+      });
+}
+
+/**
+ * Mark the end of a buffer read operation for an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ */
+static void buffer_op_stop_read(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops = bitfield_set_field32(
+      usbdev->active_buffer_ops,
+      (bitfield_field32_t){
+          .mask = 1, .index = USBDEV_READ_BUFFER_OP_BIT_INDEX, .value = 0,
+      });
+}
+
+/**
+ * Check if there is an active buffer read operation for an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ * @return `true` if there is an active buffer read operation for the endpoint,
+ * `false` otherwise.
+ */
+static bool buffer_op_is_reading(dif_usbdev_t *usbdev) {
+  return bitfield_get_field32(
+             usbdev->active_buffer_ops,
+             (bitfield_field32_t){
+                 .mask = 1, .index = USBDEV_READ_BUFFER_OP_BIT_INDEX,
+             }) > 0;
+}
+
+/**
+ * Utility functions
+ */
+
+/**
+ * Checks if the given value is a valid `dif_usbdev_toggle_t` variant.
+ */
+static bool is_valid_toggle(dif_usbdev_toggle_t val) {
+  return val == kDifUsbdevToggleEnable || val == kDifUsbdevToggleDisable;
+}
+
+/**
+ * Checks if the given value is a valid `dif_usbdev_power_sense_override_t`
+ * variant.
+ */
+static bool is_valid_power_sense_override(
+    dif_usbdev_power_sense_override_t val) {
+  return val == kDifUsbdevPowerSenseOverrideDisabled ||
+         val == kDifUsbdevPowerSenseOverridePresent ||
+         val == kDifUsbdevPowerSenseOverrideNotPresent;
```

Comment:
For something like this, I would personally have written it as a switch, which feels more consise... but up to you.

Reply:
> but up to you.

I try to avoid fallthrough in switch statements whenever I can :)